roadmap/implementers-guide/src/runtime/router.md

```diff
@@ -158,6 +161,14 @@ The following routines are intended to be invoked by paras' upward messages.
 
 Candidate Acceptance Function:
 
+* `check_upward_messages(P: ParaId, Vec<UpwardMessage>`:
+  1. Checks that there are at most `config.max_upward_msg_num_per_candidate` messages.
+  1. Checks each upward message individually depending on its kind:
+  1. If the message kind is `Dispatchable`:
+      1. Check that `data` actually decodes to a valid entrypoint / dispatchable
+      1. Check that the entrypoint weight doesn't exceed `config.max_parachain_ump_dispatch_weight`
```

Comment:
so this weight is per-message and not total in queue?

Reply:
given the validation data refactoring, i wonder how parachain block production functions are supposed to know that they are not producing too heavy of calls.

Maybe it would make sense to put a weight matrix of some kind into the `NonPersistedValidationData`. However the exact format doesn't seem obvious, given relay-chain runtime upgrades, for instance. relay-chain runtime upgrades can occur at any block and can change the weight. This may pose a barrier for contextual execution.

Reply:
that's right.

Reply:
yeah, the problem is that such a matrix can be invalidated on a chain upgrade anyway, i.e. actual weights could change. (although not sure if it will be a big of a deal in practice).

Can we delay upgrades to the session boundary?

Reply:
1. We want the acceptance criteria to be pre-evaluable by parachain validators only having access to `ValidationData` (the persisted and non persisted parts) for a candidate.
1. The acceptance function says that the weight of each dispatchable should not exceed a given maximum.
1. that maximum is easy to convey, we just put `max_parachain_ump_dispatch_weight` into `NonPersistedValidationData`.
1. the "the weight of each dispatchable" part is tricky. That implies (a) the need to know how to decode a dispatchable (also worth keeping in mind that we might need to duplicate this logic, if we don't go with the shared library approach) to extract the entrypoint and arguments and (b) a way to associate a dispatchable with a concrete weight cost (I guess the weight matrix you mentioned). However, some static matrix I guess won't do without further constraints because the weight is not a fixed value but rather a function of the entrypoint and its arguments.
1. The encoding and the matrix can be potentially invalidated during the upgrade.

Yeah, that's tricky.